open_spiel/games/checkers.cc

```diff
@@ -0,0 +1,546 @@
+    return {0., 0.};
+  } else if (outcome_ == Player{0}) {
+    return {1.0, -1.0};
+  } else if (outcome_ == Player{1}){
+    return {-1.0, 1.0};
+  }
+  return {0., 0.};
+}
+
+std::string CheckersState::InformationStateString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return HistoryString();
+}
+
+std::string CheckersState::ObservationString(Player player) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+  return ToString();
+}
+
+void CheckersState::ObservationTensor(Player player,
+                                     absl::Span<float> values) const {
+  SPIEL_CHECK_GE(player, 0);
+  SPIEL_CHECK_LT(player, num_players_);
+
+  TensorView<3> view(values, {kCellStates, rows_, columns_},
+                               true);
+
+  // Observation Tensor Representation:
+  //   Plane 0: 1's where the current player's pieces are, 0's elsewhere.
+  //   Plane 1: 1's where the oppponent's pieces are, 0's elsewhere.
+  //   Plane 2: 1's where the current player's crowned pieces are, 0's elsewhere.
+  //   Plane 3: 1's where the oppponent's crowned pieces are, 0's elsewhere.
+  //   Plane 4: 1's where the empty cells are, 0's elsewhere.
+  for (int row = 0; row < rows_; row++) {
+    for (int column = 0; column < columns_; column++) {
+      int plane = ObservationPlane(BoardAt(row, column), player);
+      view[{plane, row, column}] = 1.0;
+    }
+  }
+}
+
+void CheckersState::UndoAction(Player player, Action action) {
+  std::vector<int> values =
+      UnrankActionMixedBase(action, {rows_, columns_, kNumDirections, kNumMoveType, kNumPieceType, kNumPieceType});
+
+  const int start_row = values[0];
+  const int start_column = values[1];
+  const int direction = values[2];
+  const int move_type = values[3];
+  const int captured_piece_type = values[4];
+  const int player_piece_type = values[5];
+  
+  current_player_ = player;
+  outcome_ = kInvalidPlayer;
+  move_number_--;
+
+  int end_row, end_column;
+  CellState player_piece = player_piece_type == PieceType::kMan ? PlayerToState(player) : CrownState(PlayerToState(player));
+
+  switch (move_type) {
+    case MoveType::kNormal:
+      end_row = start_row + kDirRowOffsets[direction];
+      end_column = start_column + kDirColumnOffsets[direction];
+      SetBoard(start_row, start_column, player_piece);
+      SetBoard(end_row, end_column, CellState::kEmpty);
+      break;
+    case MoveType::kCapture:
+      end_row = start_row + kDirRowOffsets[direction] * 2;
+      end_column = start_column + kDirColumnOffsets[direction] * 2;
+      SetBoard(start_row, start_column, player_piece);
+      SetBoard(end_row, end_column, CellState::kEmpty);
+      CellState captured_piece = OpponentState(PlayerToState(player));
+      SetBoard((start_row + end_row) / 2, (start_column + end_column) / 2, 
+        captured_piece_type == PieceType::kMan ? captured_piece : CrownState(captured_piece));
```

Comment:
This wrapping is a bit odd.

Reply:
Have introduced a function here to improve readability